File: src/main/java/com/google/android/finsky/api/model/Document.java
package com.google.android.finsky.api.model;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.SparseArray;

import com.google.android.finsky.protos.nano.Messages.AppDetails;
import com.google.android.finsky.protos.nano.Messages.Common;
import com.google.android.finsky.protos.nano.Messages.DocV2;

import java.util.ArrayList;
import java.util.List;

public class Document {

    private final DocV2 doc;
    private SparseArray<List<Common.Image>> mImageTypeMap;

    public Document(@NonNull DocV2 docV2) {
        this.doc = docV2;
    }

    public AppDetails getAppDetails() {
        if (this.doc.details != null)
        {
            return this.doc.details.appDetails;
        }
        return new AppDetails();
    }

    public String getTitle() {
        return this.doc.title;
    }

    public String getDocId() {
        return this.doc.docid == null ? "" : this.doc.docid;
    }

    public int getBackend() {
        return this.doc.backendId;
    }

    public String getDetailsUrl() {
        return this.doc.detailsUrl == null ? "" : this.doc.detailsUrl;
    }

    public String getCreator() {
        return this.doc.creator == null ? "" : this.doc.creator;
    }

    public Common.Offer getOffer() {
        Common.Offer offer =  this.getOffer(Common.Offer.TYPE_1); // Type 1 ?
        if (offer == null)
        {
            offer = new Common.Offer();
        }
        return offer;
    }

    public @Nullable Common.Offer getOffer(final int type) {
        for (final Common.Offer offer : this.doc.offer) {
            if (offer.offerType == type) {
                return offer;
            }
        }
        return null;
    }

    public @Nullable String getIconUrl() {
        List<Common.Image> images = this.getImageTypeMap().get(4);
        if (images == null) {
            return null;
        }
        if (images.size() > 0) {
            return images.get(0).imageUrl;
        }
        return null;
    }

    private SparseArray<List<Common.Image>> getImageTypeMap() {
        if (this.mImageTypeMap == null) {
            this.mImageTypeMap = new SparseArray<>();
            for (final Common.Image image2 : this.doc.image) {
                final int imageType = image2.imageType;
                if (this.mImageTypeMap.get(imageType) == null) {
                    this.mImageTypeMap.put(imageType, new ArrayList<Common.Image>());
                }
                this.mImageTypeMap.get(imageType).add(image2);
            }
        }
        return this.mImageTypeMap;
    }
}

File: playstore/src/main/java/finsky/api/model/Document.kt
package finsky.api.model

import android.util.SparseArray
import finsky.protos.*

class Document(private val doc: DocV2) {

    val appDetails: AppDetails
        get() = this.doc.details.appDetails

    val title: String
        get() = this.doc.title

    val docId: String
        get() = if (this.doc.docid == null) "" else this.doc.docid

    val backend: Int
        get() = this.doc.backendId

    val detailsUrl: String
        get() = if (this.doc.detailsUrl == null) "" else this.doc.detailsUrl

    val creator: String
        get() = if (this.doc.creator == null) "" else this.doc.creator

    // Type 1 ?
    val offer: Common.Offer
        get() {
            var offer = this.doc.offerList.firstOrNull { it.offerType == Common.Offer.TYPE.TYPE_1_VALUE }
            if (offer == null) {
                offer = Common.Offer.getDefaultInstance()
            }
            return offer!!
        }

    val iconUrl: String?
        get() {
            val images = this.imageTypeMap.get(4) ?: return null
            return if (images.isNotEmpty()) {
                images[0].imageUrl
            } else null
        }

    private val imageTypeMap: SparseArray<MutableList<Common.Image>> by lazy {
        val typeMap = SparseArray<MutableList<Common.Image>>()
        for (image2 in this.doc.imageList) {
            val imageType = image2.imageType
            if (typeMap.get(imageType) == null) {
                typeMap.put(imageType, mutableListOf())
            }
            typeMap.get(imageType).add(image2)
        }
        typeMap
    }
}
